Add an option to skip integrated BCs when the variable is not defined next to the boundary

GiudGiud opened the pull request:
closes #29360

@jmeier 

Came back to it since it may help https://github.com/idaholab/moose/discussions/29699 as well

GiudGiud committed:
Add an option to skip integrated BCs when the variable is not defined
next to the boundary
closes #29360

GiudGiud committed:
Add a test for new skip-integrated-bc when variable is no longer defined

GiudGiud committed:
Add helpful documentation for mesh modifiers
refs #29360

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2640906/), step Docs: sync website on 7185678 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29700/site)

This comment will be updated on new commits.

GiudGiud commented:
Looks like this triggers errors on both the HFEM and mortar BCs. I will be a little delayed in investigating this.

lindsayad commented:
looks like some associated test failures

GiudGiud commented:
Good catch. Pretty unlucky, but recovering post a subdomain change can cause the old subdomain to be missing!

GiudGiud commented:
otherwise this should be ready I think

GiudGiud commented:
for real this time

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2678803/), step Generate coverage on cbf913f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>93913d</th>
      <th colspan=4>#29700 cbf913</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.28%</td>
      <td>85.28%</td>
      <td>+0.00%</td>
      <td>85.71%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108338</td>
      <td>108351</td>
      <td>+13</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18697</td>
      <td>18696</td>
      <td>-1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29700/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29700/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29700/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.71% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Test failures are not related. An openMPI stochastic failure and a build failure on mac

lindsayad commented on framework/src/bcs/IntegratedBC.C:
```diff
@@ -85,6 +95,16 @@ IntegratedBC::computeResidual()
+  if (_local_re.empty())
+  {
+    if (!_skip_execution_outside_variable_domain)
+      paramError("skip_execution_outside_variable_domain",
+                 "This boundary condition is being executed outside the domain of definition of "
+                 "its variable");
```
IIRC `paramError` is not actually thread-safe

GiudGiud replied:
are you sure because it seems to forward to this, which has a lock in it so someone thought about thread safety

```

[[noreturn]] void
mooseErrorRaw(std::string msg, const std::string prefix)
{
  if (Moose::_throw_on_error)
    throw std::runtime_error(msg);

  msg = mooseMsgFmt(msg, "*** ERROR ***", COLOR_RED);

  std::ostringstream oss;
  oss << msg << "\n";

  // this independent flush of the partial error message (i.e. without the
  // trace) is here because trace retrieval can be slow in some
  // circumstances, and we want to get the error message out ASAP.
  msg = oss.str();
  if (!prefix.empty())
    MooseUtils::indentMessage(prefix, msg);
  {
    Threads::spin_mutex::scoped_lock lock(moose_stream_lock);
    Moose::err << msg << std::flush;
  }
```

lindsayad replied:
If you add a test for the error, I will check

GiudGiud replied:
ok did that, and turned on the clang thread sanitizer recipe on civet so hopefully you dont need to check manually?

lindsayad replied:
I get a ghosted index OOB with 4 threads

GiudGiud replied:
no luck on that recipe

lindsayad replied:
that's only an error in cleanup, that's probably fine

GiudGiud replied:
Let's raise an issue. paramError should be thread safe. The additional context it gives is valuable even inside threaded loops

lindsayad replied:
I think we should resolve the issue with `--n-threads=4` before continuing. You get OOB indexing whether you are running with `BCs/neum/skip_execution_outside_variable_domain` `true` or `false`

GiudGiud replied:
can you paste the OOB for that? maybe it s obvious

lindsayad replied:
```
Time Step 2, time = 2, dt = 1

We caught a libMesh error in ThreadedElementLoopBase:No index 8 in ghosted vector.
Vector contains [0,6)
And empty ghost array.

Stack frames: 16
0: ___interceptor_backtrace
1: libMesh::print_trace(std::ostream&)
2: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
3: libMesh::PetscVector<double>::map_global_to_local_index(unsigned long) const
4: libMesh::PetscVector<double>::get(std::vector<unsigned long, std::allocator<unsigned long> > const&, double*) const
5: MooseVariableDataBase<double>::fetchDoFValues()
6: MooseVariableData<double>::computeValues()
7: MooseVariableFE<double>::computeElemValuesFace()
8: SystemBase::reinitElemFace(libMesh::Elem const*, unsigned int, unsigned int)
9: FEProblemBase::reinitElemFace(libMesh::Elem const*, unsigned int, unsigned int)
10: ComputeMaterialsObjectThread::onBoundary(libMesh::Elem const*, unsigned int, short, libMesh::Elem const*)
11: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
12: void* libMesh::Threads::run_body<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeMaterialsObjectThread>(void*)
13: ../../../moose_test-devel(+0x62833) [0x55f9012a8833]
14: /lib/x86_64-linux-gnu/libc.so.6(+0x9ca94) [0x7f90bf017a94]
15: __clone
[0] /data/lindad/projects/thread-debugging/scripts/../libmesh/installed/include/libmesh/petsc_vector.h, line 1080, compiled Feb 11 2025 at 13:26:10
```

GiudGiud replied:
ok so materials are being computed where they should not. I guess that's expected

GiudGiud replied:
and my fix only catches the test. I ll add an error

GiudGiud replied:
should be fine now. Though there s no real guarantees that it would not happen again with:
- another system that also pre-computes things and that also checks for boundaries existing before pre-computing gets in the same bad re-init

GiudGiud committed:
Address Alex's review:
- add a test for the error

GiudGiud committed:
Address Alex's review:
- prevent OOB access when initializing stateful material properties on boundary

lindsayad commented:
geomsearch failure looks real

moosebuild commented:
Job [Framework 2](https://civet.inl.gov/job/2684585/) on 0129013 : invalidated by @GiudGiud

I think it s unrelated

GiudGiud commented:
it passed fine

lindsayad commented:
It failed for me locally too when I tested it yesterday. Hmm

lindsayad commented:
It passes valgrind so that's good

GiudGiud commented:
well valgrind is not threaded and I fixed the combination we were seeing not working. 
I ll add a "this is an experimental feature" in the doco in the next fixup PR

GiudGiud commented:
well valgrind is not threaded and I fixed the combination we were seeing not working. 
I ll add a "this is an experimental feature" in the doco in the next fixup PR

lindsayad commented:
Now that we're down to it ... are there better ways to do this? `BoundaryCondition` has a `shouldApply` `virtual`. Should we override that in `IntegratedBC` to check whether the variable has blocks on `_current_elem->subdomain_id()`, or combine that with your parameter? Relying on `_local_re` or `_local_ke` feels dangerous to me. I don't think there's always going to be a 1-to-1 relationship between those being empty and variables undergoing subdomain activation/deactivation. I've written plenty of buggy code that leads to those variables being empty that have nothing to do with mesh modification

GiudGiud commented:
seems like a possible rework. I am not a fan of no arguments + rely on attributes being set.
look at these caveats on shouldApply already:

```
   * It is not safe to use variable values in this function, since (a) this is not called inside a
   * quadrature loop,
   * (b) reinit() is not called, thus the variables values are not computed.
   * NOTE: In NodalBC-derived classes, we can use the variable values, since renitNodeFace() was
   * called before calling
   * this method. However, one has to index into the values manually, i.e. not using _qp.
  ```

>  I've written plenty of buggy code that leads to those variables being empty that have nothing to do with mesh modification

seems like a perk to now have the check we have now then. and you can use the new boolean to make your buggy code not contribute to the residual and Jacobian :D

lindsayad commented:
> I am not a fan of no arguments + rely on attributes being set.

Isn't this exactly the state of `computeResidual()` 😄 

> I am not a fan of no arguments + rely on attributes being set.

None of those caveats apply to whether a variable is defined on a block

> and you can use the new boolean to make your buggy code not contribute to the residual and Jacobian :D

right, definitely not a perk 😄 Failed assertions around those variables being empty should not be easily avoided by setting a parameter that may not be applicable to the failure case. I admit it's unlikely that a user would deliberately work around an assertion that way. I just think that finding these variables to be empty should not be a decision-tree point. It should always denote something wrong

lindsayad commented:
And that geomsearch test is still failing for me locally, 10/10 times with `--n-threads=2`

lindsayad commented:
Oh I know why it fails for me so consistently now ... it's because that test always has failed point inversions which were previously thread unsafe and TSAN exited with a non-zero exit code. With https://github.com/libMesh/libmesh/pull/4080 that test now has a zero exit code

GiudGiud commented:
alright I ll look into shouldApply and add to the caveat list

GiudGiud commented:
it's "public" too. Going to have to trust the devs

lindsayad commented:
I don't see why we'd need to trust them any more than before

lindsayad commented:
Fundamentally this PR is about whether we should apply a boundary condition, isn't it?

GiudGiud commented:
it is. 

but checking the block restriction of every the variable every single time we use a BC is not a good idea for performance so I stayed away from doing that, which is what should be done in shouldApply(). With more thinking, I can do this a little better than what I thought

> wasting your time

I dont think reviews are a waste of time. This PR will come out a lot better, you'll see

lindsayad commented:
I just meant it would have been better to raise it up front

GiudGiud commented:
such is life

GiudGiud committed:
Use shouldApply() instead of adding logic within the computeResidual/Jacobian to detect the out-of-domain
_ this means shouldApply() now needs the _current_elem to be set

GiudGiud committed:
Make shouldApply const
it's being called in threaded regions, easier to have it const

GiudGiud commented:
nice I m getting line changes in the Richards module. peak MOOSE iron age module

lindsayad commented:
You did make me worried about the cost of querying a set for every element for every boundary. So maybe we should cache a member variable set in a `subdomainSetup()` override?

GiudGiud commented:
we are going to have to I think. My guess is that the test failures are coming from checking _current_elem in shouldApply() when those objects are on _current_side_elem.
Side note but, whether it's the problem or not, this is why I don't like relying on updating attributes in the back

lindsayad commented:
Your previous code was relying on updating `_local_re` or `_local_ke` in the back. I don't see a qualitative difference

lindsayad commented:
> we are going to have to I think. My guess is that the test failures are coming from checking _current_elem in shouldApply() when those objects are on _current_side_elem.

The objects would not be associated with `_current_side_elem`. Integrated BCs are always associated with an element and a side, not the lower-dimensional element on the element side

GiudGiud committed:
Only check the variable domain of definition if in the "skip when subdomain is not part of variable block restriction"
option or if in debug mode

This limits any performance lost to this option.
This also means we do not, as before this PR, detect when the bad situation is happening
in opt mode

Update an assert in HDG
Allow again using stateful matprops since the reinit done on the boundary with missing values should
be calling shouldApply as expected

GiudGiud commented:
> Your previous code was relying on updating _local_re or _local_ke in the back. I don't see a qualitative difference

right next to code that uses those two attributes. if they were not set, something was wrong.

GiudGiud commented:
`subdomainSetup()` is not called on integrated BCs. And even if it were, BCs are not sorted by blocks only by boundaries so they dont register as active block objects.
A new `boundarySetup()` call would be a lot of overhead for just this one functionality and does not fit well in the current residual loops (on elements, then on sides)

GiudGiud commented:
I think maybe we should do something in 
```
void
NonlinearThread::subdomainChanged()
```
instead

GiudGiud commented:
this should be acceptable?

GiudGiud commented:
This new option is still only working for the residual/jacobian loops.
If we need it working for `ComputeMaterialsObjectThread` for example, we need to have the same change in SubdomainChanged()

and if the very last thing in the residual loop is skipping, then this BC will no longer show as "shouldApply" for any other loop!

GiudGiud commented:
no can do in the problem.
At this point I m going to say that if you want this functionality, you can override shouldApply in your BCs to check the variable subdomain.

GiudGiud commented:
ok this should be reasonable. 
- no more fiddling in the back end between subdomain and boundaries in the problem
- only expecting a performance hit when the special mode in on or in debug mode

lindsayad commented on framework/src/bcs/IntegratedBC.C:
```diff
@@ -15,6 +15,7 @@
 #include "SystemBase.h"
 #include "MooseVariableFE.h"
 #include "MooseVariableScalar.h"
+#include "MaterialPropertyStorage.h"
```
is this needed?

GiudGiud replied:
not anymore. will clear

lindsayad commented on framework/src/bcs/IntegratedBCBase.C:
```diff
@@ -61,3 +72,27 @@ IntegratedBCBase::prepareShapes(const unsigned int var_num)
+bool
+IntegratedBCBase::shouldApply() const
+{
+#ifndef NDEBUG
```
```suggestion
#ifdef DEBUG
```
We typically add O(1) assertion like checks for devel and restrict checks that depend on N (N in this case the number of subdomains) to debug. Admittedly it's pretty rare to have a lot of subdomains

GiudGiud replied:
I think I have it right right now
https://github.com/microsoft/qsharp-runtime/issues/600

NDEBUG should be defined. DEBUG may be defined

GiudGiud replied:
oh nevermind I see what you mean

GiudGiud replied:
I think searching in a set is log(n) so should be fine to have in devel

GiudGiud replied:
same I use devel mostly. I can go either way. if you find it slow sure let s change it to debug

lindsayad replied:
probably not worth the time we're spending over it

GiudGiud replied:
sure. I tried `#ifdef DEBUG` and I am not seeing it trigger in devel

also this does not work
```
    method = 'dbg|devel'
```

I ll leave it as is

GiudGiud replied:
ooooh

lindsayad commented on framework/src/meshmodifiers/ElementSubdomainModifierBase.C:
```diff
@@ -215,11 +216,13 @@ ElementSubdomainModifierBase::modify(
 
   // Initialize solution and stateful material properties
   applyIC(/*displaced=*/false);
+  // if (_fe_problem.getMaterialWarehouse().hasActiveObjects(0))
```
```suggestion
```

GiudGiud replied:
oh this is an accident. this line should be there and uncommented

lindsayad commented on framework/src/meshmodifiers/ElementSubdomainModifierBase.C:
```diff
@@ -215,11 +216,13 @@ ElementSubdomainModifierBase::modify(
   if (_displaced_mesh)
   {
     applyIC(/*displaced=*/true);
+    // if (_fe_problem.getMaterialWarehouse().hasActiveObjects(0))
```
```suggestion
```

lindsayad reviewed (changes requested):
I like this version better for sure

GiudGiud committed:
Make variable domain definition check on integrated BCs part of devel rather than just debug

GiudGiud commented:
patch for SAM on const shouldApply

https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/1015
